### PR TITLE
fix(reports): force data generation in csv reports

### LINKED ISSUE #22189: [reports] GET chart/<pk>/data api doesn't respect force parameter
Api to get data of a saved chart used by reports to send csv data doesn't respect force parameter resulting in reports having stale data.

#### How to reproduce the bug

1. setup caching for data and report
2. create table chart, so it caches the data
3. change the data in the datasource without busrting cache of the chart
4. create report with csv and enable `Ignore cache when generating screenshot`
5. see the report having stale data.

### Expected results
As `Ignore cache when generating screenshot` is enabled, report should have fresh data from datasource.
Or this option is only for reports with screenshot as its name suggests?

### Actual results
Report has stale data.

### Environment

(please complete the following information):

- browser type and version:
- superset version: 1.5.1 and 2.0
- python version: `python --version`
- node.js version: `node -v`
- any feature flags active: ALERT_REPORTS

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [ ] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [ ] I have reproduced the issue with at least the latest released version of superset.
- [ ] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by mayurnewase**
### SUMMARY
fixes: #22189 

### TESTING INSTRUCTIONS

1. create table chart
2. setup csv report on that chart and enable `Ignore cache when generating screenshot`
3. update data in datasource without bursting cache
4. check csv in the report received, should contain new data

### ADDITIONAL INFORMATION

- [x] Has associated issue: #22189 
- [x] Required feature flags: `ALERTS_REPORTS`
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by mayurnewase** [comment]
Should we also change `Ignore cache when generating screenshot` to `Ignore cache when generating report` as CSV, json do not need screenshots.

File: superset/charts/data/api.py [hunk 1]
```diff
@@ -130,6 +135,7 @@ def get_data(self, pk: int) -> Response:
             "format", ChartDataResultFormat.JSON
         )
         json_body["result_type"] = request.args.get("type", ChartDataResultType.FULL)
+        json_body["force"] = request.args.get("force", False)
```
Comment (zhaoyongjie) [inline superset/charts/data/api.py:138]: In order to handle `false` as a string in the force, an explicit type conversion should be in here. for example:
```type
        json_body["force"] = explict_bool_conversion(request.args.get("force"))
```

Comment (mayurnewase) [reply superset/charts/data/api.py:138]: umm I think that is taken care by `_create_query_context_from_form` function below using mashmallow schema's load function.
I tested with false too and it's working as expected, do you still need this conversion here?

Comment (zhaoyongjie) [reply superset/charts/data/api.py:138]: if exists another deserialization helper on the json_body, the default value of `request.args.get("force")` should be None instead of `request.args.get("force", False)`, because the `request.args.get("force")` will always return `Optional[str]`.

File: tests/integration_tests/charts/data/api_tests.py [hunk 2]
```diff
@@ -862,6 +862,59 @@ def test_chart_data_get(self):
...
+
+        # wrapping original class, so we can test output too.
+        with mock.patch(
+            "superset.charts.data.api.ChartDataCommand", wraps=ChartDataCommand
+        ) as command:
+            rv = self.get_assert_metric(
+                f"api/v1/chart/{chart.id}/data/?force=true", "get_data"
+            )
+            data = json.loads(rv.data.decode("utf-8"))
+            query_context = command.call_args.args[0]
```
[the remark this answers is not in the document]
Comment (mayurnewase) [reply tests/integration_tests/charts/data/api_tests.py:911]: so you suggesting, we just check the output and not the force param in query_context?

File: superset/charts/data/api.py (hunk 1 above)
Comment (mayurnewase) [reply superset/charts/data/api.py]: ok updated.

Comment (mayurnewase) [reply superset/charts/data/api.py]: if we set force to be `None` then masrhmallow raises error saying it should be a boolean.

Comment (zhaoyongjie) [reply superset/charts/data/api.py]: add `allow_none` argument in the rule. 
```
    force = fields.Boolean(
        description="Should the queries be forced to load from the source. "
        "Default: `false`",
        allow_none=True
    )
```

File: tests/integration_tests/charts/data/api_tests.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply tests/integration_tests/charts/data/api_tests.py:911]: we should verify the `is_cached` field in the response.

```Python
    test_client.post(f"api/v1/chart/{chart.id}/data/?force=true")
    # should through cache
    rv = test_client.post(f"api/v1/chart/{chart.id}/data/?force=true")
    assert rv.json["result"][0]["is_cached"] is None

    test_client.post(f"api/v1/chart/{chart.id}/data")
    # should get response from the cache
    rv = test_client.post(f"api/v1/chart/{chart.id}/data/?force=true")
    assert rv.json["result"][0]["is_cached"] is not None
```

[the remark this answers is not in the document]
Comment (mayurnewase) [reply tests/integration_tests/charts/data/api_tests.py:911]: yeah good call, updating.

[the remark this answers is not in the document]
Comment (mayurnewase) [reply tests/integration_tests/charts/data/api_tests.py:911]: I wasn't sure if redis runs while integrations tests locally and on CI.

[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply tests/integration_tests/charts/data/api_tests.py:911]: yes, the [DATA_CACHE_CONFIG](https://github.com/apache/superset/blob/eba7b3d074350c3429778259baef19b7995f60ae/tests/integration_tests/superset_test_config.py#L106) has been set on the integrations test.

[the remark this answers is not in the document]
Comment (mayurnewase) [reply tests/integration_tests/charts/data/api_tests.py:911]: updated.

File: tests/integration_tests/charts/data/api_tests.py [hunk 3]
```diff
@@ -862,6 +862,63 @@ def test_chart_data_get(self):
...
+            }
+        )
+
+        self.get_assert_metric(f"api/v1/chart/{chart.id}/data/", "get_data")
+
+        rv = self.get_assert_metric(
+            f"api/v1/chart/{chart.id}/data/?force=true", "get_data"
+        )
+
+        data = json.loads(rv.data.decode("utf-8"))
```
Comment (zhaoyongjie) [inline tests/integration_tests/charts/data/api_tests.py:909]: These codes are unrelated to this test case, should we remove these?

File: tests/integration_tests/charts/data/api_tests.py [hunk 4]
```diff
@@ -862,6 +862,63 @@ def test_chart_data_get(self):
...
+        assert rv.json["result"][0]["is_cached"] is None
+        assert rv.mimetype == "application/json"
+        assert data["result"][0]["status"] == "success"
+        assert data["result"][0]["rowcount"] == 2
```
Comment (zhaoyongjie) [inline tests/integration_tests/charts/data/api_tests.py:913]: these codes are redundant as well.

File: tests/integration_tests/charts/data/api_tests.py [hunk 5]
```diff
@@ -862,6 +862,63 @@ def test_chart_data_get(self):
...
+
+        rv = self.get_assert_metric(f"api/v1/chart/{chart.id}/data/", "get_data")
+        data = json.loads(rv.data.decode("utf-8"))
```
Comment (zhaoyongjie) [inline tests/integration_tests/charts/data/api_tests.py:916]: and these

File: tests/integration_tests/charts/data/api_tests.py [hunk 6]
```diff
@@ -862,6 +862,63 @@ def test_chart_data_get(self):
...
+        assert rv.json["result"][0]["is_cached"] is not None
+        assert rv.mimetype == "application/json"
+        assert data["result"][0]["status"] == "success"
+        assert data["result"][0]["rowcount"] == 2
```
Comment (zhaoyongjie) [inline tests/integration_tests/charts/data/api_tests.py:920]: these

File: tests/integration_tests/charts/data/api_tests.py (hunk 3 above)
Comment (mayurnewase) [reply tests/integration_tests/charts/data/api_tests.py:909]: so instead of raw `test_client`, this file uses `get/post_assert_metric`, so just wanted to be consistent.

Comment (mayurnewase) [reply tests/integration_tests/charts/data/api_tests.py:909]: can remove other data asserts.

**Review by zhaoyongjie — APPROVED** [review]
Thanks for your patience! LGTM.

**Comment by mayurnewase** [comment]
Thanks for the reviews 😀
